plone/formwidget/namedfile/utils.py

```diff
@@ -0,0 +1,46 @@
+# -*- coding: utf-8 -*-
+from datetime import datetime
+from datetime import timedelta
+from plone.formwidget.namedfile.interfaces import IFileUploadMap
+from zope.annotation.interfaces import IAnnotations
+from zope.component import adapter
+from zope.interface import implementer
+from zope.interface import Interface
+from ZPublisher.HTTPRequest import FileUpload
+from BTrees.IOBTree import IOBTree
+from random import randint
+
+FILE_UPLOAD_MAP_KEY = 'file_upload_map'
+FILE_UPLOAD_EXPIRATION_TIME = 30*60  # seconds
+
+
+def is_file_upload(item):
+    """Check if ``item`` is a file upload.
+    """
+    return isinstance(item, FileUpload)
+
+
+@implementer(IFileUploadMap)
+@adapter(Interface)
+class FileUploadMap(object):
+    """Temporary storage adapter for file uploads.
+    To be used to not need to re-upload files after form submission errors.
+    """
+
+    def __init__(self, context):
+        self.context = context
+
+    @property
+    def upload_map(self):
+        annotations = IAnnotations(self.context)
+        upload_map = annotations.get(FILE_UPLOAD_MAP_KEY, IOBTree())
```

Comment:
@jensens: when I change this to a plain dict ``{}``, this works exactly the same way. Should we still keep IOBTree?

Reply:
a dict is written as a whole in ZODB, a Btree is divided into buckets, so yes for a tree and no, its not the same.